kernel/arch/dreamcast/include/dc/maple/controller.h

```diff
@@ -24,149 +27,466 @@
+
+    \param addr             Maple BUS address to poll for the button mask
+                            on, or 0 for all ports.
+    \param btns             Mask of all buttons which should be pressed to
+                            trigger the callback.
+
+    \sa cont_btn_callback
 */
-typedef void (*cont_btn_callback_t)(uint8 addr, uint32 btns);
+typedef void (*cont_btn_callback_t)(uint8_t addr, uint32_t btns);
 
-/** \brief  Set an automatic button press callback.
+/** \brief   Set an automatic button press callback.
+    \ingroup controller_inputs
 
     This function sets a callback function to be called when the specified
     controller has the set of buttons given pressed.
 
-    \param  addr            The controller to listen on. This value can be
-                            obtained by using maple_addr().
+    \note 
+    The callback gets invoked for the given maple port; however, providing
+    an address of '0' will cause it to be invoked for any port with a 
+    device pressing the given buttons. Since you are passed back the address 
+    of this device, You are free to implement your own filtering logic within
+    your callback.
+
+    \warning
+    The provided callback function is invoked within a context which has
+    interrupts disabled. This means that you should not do any sort of complex
+    processing or make any API calls which depend on interrupts to complete,
+    such as Maple or etherent processing whichy rely on packet transmission,
+    any sleeping or threading calls, blocking on any sort of file I/O, etc. 
+    This mechanism is typically used to quickly terminate the application
+    and should be used with caution.
+
+    \param  addr            The controller to listen on (or 0 for all ports). 
+                            This value can be obtained by using maple_addr().
     \param  btns            The buttons bitmask to match.
     \param  cb              The callback to call when the buttons are pressed.
 */
-void cont_btn_callback(uint8 addr, uint32 btns, cont_btn_callback_t cb);
+void cont_btn_callback(uint8_t addr, uint32_t btns, cont_btn_callback_t cb);
+
+/** \defgroup controller_query_caps Querying Capabilities
+    \brief    API used to query for a controller's capabilities
+    \ingroup  controller
+    
+    The following API is used to query for the support of individual 
+    or groups of capabilities by a particular device.  
+*/
 
-/** \defgroup   controller_caps Controller capability bits.
+/** \defgroup controller_caps Capabilities
+    \brief    Bit masks used ot identify controller capabilities
+    \ingroup  controller_query_caps
 
     These bits will be set in the function_data for the controller's deviceinfo
-    if the controller supports the corresponding feature.
+    if the controller supports the corresponding button/axis capability.
+
+    @{
+*/
+#define CONT_CAPABILITY_C               (1<<24)     /**< \brief C button capability mask. */
```

Comment:
Might be worthwhile to note that the strange ordering is to match the above controller_inputs defgroup